src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/ServerVariables.cs

```diff
@@ -19,44 +26,75 @@ public static PatternSegment FindServerVariable(string serverVariable, ParserCon
+                    break;
                 case "HTTP_HOST":
-                    return new HeaderSegment(HeaderNames.Host);
+                    patternSegment = new HeaderSegment(HeaderNames.Host);
+                    break;
                 case "HTTP_REFERER":
-                    return new HeaderSegment(HeaderNames.Referer);
+                    patternSegment = new HeaderSegment(HeaderNames.Referer);
+                    break;
                 case "HTTP_USER_AGENT":
-                    return new HeaderSegment(HeaderNames.UserAgent);
+                    patternSegment = new HeaderSegment(HeaderNames.UserAgent);
+                    break;
                 case "HTTP_CONNECTION":
-                    return new HeaderSegment(HeaderNames.Connection);
+                    patternSegment = new HeaderSegment(HeaderNames.Connection);
+                    break;
                 case "HTTP_URL":
-                    return new UrlSegment();
+                    patternSegment = new UrlSegment();
+                    break;
                 case "HTTPS":
-                    return new IsHttpsUrlSegment();
+                    patternSegment = new IsHttpsUrlSegment();
+                    break;
                 case "LOCAL_ADDR":
-                    return new LocalAddressSegment();
+                    patternSegment = new LocalAddressSegment();
+                    break;
                 case "HTTP_PROXY_CONNECTION":
                     throw new NotSupportedException(Resources.FormatError_UnsupportedServerVariable(serverVariable));
                 case "QUERY_STRING":
-                    return new QueryStringSegment();
+                    patternSegment = new QueryStringSegment();
+                    break;
                 case "REMOTE_ADDR":
-                    return new RemoteAddressSegment();
+                    patternSegment = new RemoteAddressSegment();
+                    break;
                 case "REMOTE_HOST":
                     throw new NotSupportedException(Resources.FormatError_UnsupportedServerVariable(serverVariable));
                 case "REMOTE_PORT":
-                    return new RemotePortSegment();
+                    patternSegment = new RemotePortSegment();
+                    break;
                 case "REQUEST_FILENAME":
-                    return new RequestFileNameSegment();
+                    patternSegment = new RequestFileNameSegment();
+                    break;
                 case "REQUEST_URI":
-                    return new UrlSegment();
+                    patternSegment = new UrlSegment();
+                    break;
                 default:
                     throw new FormatException(Resources.FormatError_InputParserUnrecognizedParameter(serverVariable, context.Index));
             }
+
+            return new ServerVariable(serverVariable, new Pattern(patternSegment), ServerVariableType.RequestHeader);
```

Comment:
need some feedback on which are not and what they are instead and then i can dial this in.